README.md

```diff
@@ -114,12 +130,18 @@ in the namespace `uavcan.register`.
 The register protocol provides a highly generic interface to vendor-specific functionality
 and configuration parameters via named registers.
 
-## Non-standard regulated data types
+## Non-standard data types
 
-Every vendor must have a dedicated root namespace.
-The root namespace should be named after the vendor's legal entity name.
+Non-standard regulated data types are contained in the root namespace `com`.
```

Comment:
can this be a suggestion? Why com? Why not edu, io, ca, etc? Should we just allow any TLD?

Reply:
I think it is critical that we enclose all vendor-specific namespaces in a top-level namespace because:

> Having a vendor namespace named after said vendor in the public regulated repo would make it impossible for the same vendor to have an unregulated namespace under the same name. This is because having an identically named namespace with different contents is currently explicitly prohibited by the spec (https://github.com/UAVCAN/pyuavcan/pull/68#issuecomment-519677295).

The name of the top-level namespace does not matter much, but it's best to keep it short because we only have 50 characters for the full data type name. Using TLDs is not really critical, but I think they suit the purpose well because they are short, familiar, and their meaning is intuitively clear. I am not sure whether we should allow more than one TLD; maybe we should at least discourage that for reasons of consistency. We can always lift the restriction to use only `com` in the future. Do you think we should lift it now?

The domain name analogy should also make it intuitively clear that the public regulated repository is a common, globally shared resource, which should be well-structured and maintained carefully. Vendors are free to use other, less formal naming patters in their non-regulated or private namespaces (much like it's possible to use any TLD in a local DNS server).

Under the above model, the fact that `uavcan` is a separate namespace, not under a TLD, is desirable because it highlights its special status.

Reply:
Using TLDs is a good way to deconflict but we should support some range of them. I'd prefer one of two conventions be established:

1. Support a small set of the most common TLDs that are 2-3 chars in length. For example the original 6 plus io : `[com, org, net, int, edu, gov, mil, io]`

2. Support all IANA TLDs. This can be supported in parsers by downloading that list from http://data.iana.org/TLD/tlds-alpha-by-domain.txt

Reply:
Emergent complexity is looming on the horizon.

The spec says that if you defined a top-level namespace in one place, then that's your entire namespace and you will have to keep it there as an atomic entity. Want to define a sub-namespace somewhere separate? Nope, you can't.

The reason is that having access to the entire namespace, a DSDL front-end (currently it's only PyDSDL) can enforce versioning rules and warn early about name conflicts. The point of aggregation could be chosen arbitrarily; it's a root namespace because it maps well onto the general protocol architecture, but it could be higher (multiple root namespaces?) or lower (sub-root namespaces?).

We may remove this restriction later, but as long as it is in place, we have to keep in mind that by allowing the use of a root namespace for the public regulated namespace library (this repo), we thereby withdraw it from the set of permissible names that vendors could use for themselves. This is why allowing any TLD from the most recent set standardized by IANA is a bad idea -- what if there is a weirdly named vendor like "MOSCOW" or idk "GOOGLE"?

Imagine now that we're living in the far future where the civilization has advanced to the point where one doesn't need to keep their root namespaces atomic to guarantee compatibility. Anybody is free to name their vendor-specific namespace after an Internet domain name, like `com.nozama`. Do we still want to provide a clear separation between public regulated types and non-regulated types? Do we still want to appease hipsters registering their websites under `.io`, or is that carefully crafted special case no longer relevant?

Under a working hypothesis that the optimal answers are "yes" and "no", we arrive at the conclusion that TLD-based regulated root namespaces may not age well. Further, as long as Specification requires atomic root namespaces, it should also prohibit usage of TLDs for unregulated root namespace names, and DSDL front-ends should refuse to process them. Seriously, there are [v0 vendors out there](https://github.com/OpenMotorDrive/flow/tree/master/dsdl/com) that are already practicing this, not knowing that it is a recipe for disaster.

Should we search for a new name for the outer namespace of vendor-specific public regulated types instead of `com`?

Should we discuss the atomic namespace requirement in-depth?

Reply:
We could consider replacing `com` with `universe`.

Reply:
`public`?

Reply:
It's a keyword in many languages and data types can be public-unregulated. Maybe just `regulated`? It is unambiguous because data types cannot be private and regulated:

![image](https://user-images.githubusercontent.com/3298404/68154896-38aff400-ff51-11e9-9fcf-08012cc6c038.png)

Reply:
Or `reg` because we care about length.

Reply:
okay. `reg` but what about name-spacing here. With the core types we have everything under `uavcan` which is unlikely to result in collisions when generated code in combined with an existing codebase. All of the proposed top-level namespaces are generic enough that we may see collisions. Should we not have something with `uavcan` in it some how?

`ucreg`
`uavcan-reg`
`uavcan.reg`

?

Reply:
I understand the name conflict argument and I thought about it too. Maybe it's best to solve this not at the infrastructure level but at the implementation level, allowing the user to specify what package or library should DSDL-generated code be wrapped in?

A quick look around reveals that `reg` conflicts with existing (albeit not very useful) things out there:

- https://crates.io/crates/reg (appears to be a placeholder)
- https://pypi.org/project/reg/
- https://www.npmjs.com/package/reg (not sure what this is but it exists)

But `regulated` seems okay:

- https://crates.io/search?q=regulated (no matches)
- https://pypi.org/search/?q=regulated (no matches)
- https://www.npmjs.com/search?q=regulated (no matches)

I agree though that it's weird to use such a generic adjective to name a software package. It makes sense for DSDL, but one scrolling through a codebase that relies on DSDL might be perplexed by things like `#include <regulated/zubax/actuator/esc_group/Status_0_1.hpp>`. Despite this, I don't have better ideas at the moment.